WebClient Streamable HTTP support

chemicL opened the pull request:
Introducing WebClient support for Streamable HTTP transport.

## Motivation and Context
The 2025-03-26 specification of MCP introduces Streamable HTTP transport that replaces SSE transport.
This change introduces more resilient primitives that support the initial, WebClient-based client transport for Streamable HTTP.

Various failure modes at either the transport or logical layer have been addressed and are now fully supported by the new `WebClientStreamableHttpTransport` class. The surrounding `McpAsyncClient` and `McpClientSession` have been modified to handle possible issues according to the adequate layer's responsibility.

### Orientation

The existing APIs are organized quite logically with regards to various responsibilities using layers. Below is the status quo description and the changes required for Streamable HTTP to be implemented.

#### Rationale for the API

The client types are layered in the following way:

```
McpAsyncClient
  McpClientSession
    McpClientTransport
      McpTransportSession
      McpTransportStream
```

Here are the distinguishing features of these APIs:

`McpAsyncClient` (and the facade, `McpSyncClient`)
  - user facing class exposing capabilities like tools, etc.
  - maintains state for the initialization (and an inner class was added to encapsulate this state)
  - maintains state for capabilities handlers, these translate to lower-level handlers at JSON-RPC level in the session
  - (change) used to have a single session, now dynamically allocates sessions upon reconnects/invalidations
  - (change) used to delegate to session to manage the transport, now it owns the transport and its lifecycle

`Mcp(Client)Session`
  - internal implementation providing JSON-RPC mapping of requests with responses
  - maintains state for the requests pending responses
  - allows defining req/resp/notification handlers and keeps a state of them
  - dispatches messages to their handlers
  - (change) used to own the transport and manage its lifecycle, now it's independent

`Mcp(Client)Transport`
  - specific implementation for the transport layer of serialized JSON-RPC
  - STDIO, SSE, Streamable HTTP
  - maintains open streams in case of streamable connections (SSE, Streamable HTTP, STDIO)
  - maintains particular state of the transport type, e.g. message endpoint in case of SSE transport
  - (new) spawns transport-specific sessions (e.g. HTTP status-driven session management from streamable HTTP)
  - (new) spawns transport-specific streams that can come and go

`McpTransportSession`
 - new abstraction over transport-driven session concept
 - allows enriching the communication with transport-level session identifier to communicate with stateful servers

`McpTransportStream`
 - new abstraction for a transport-level stream (particularly SSE stream) with message id tracking for resumability purposes

## How Has This Been Tested?
* Regular integration tests using the `everything` MCP server.
* Resilience tests validating reconnections and session invalidation handling.

## Breaking Changes
The client initialization became lazy and more error-prone - connections happen upon first request and initialization is performed implicitly if it has not been issued before.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

## Additional context
The JDK HttpClient will be implemented next.

## Remaining tasks
- [x] Builder
- [x] Javadoc
- [x] Supporting DELETE call

Related #72

chemicL commented:
The CI fails for the Stdio tests. I can't reproduce it on my M1 Mac, but will try a linux box and investigate.

chemicL commented:
There is the missing bit for issuing a DELETE upon client-side session termination, I'll investigate how it could be achieved.

I will review the spec once more and see what else is missing.

bugzyz commented:
Hi @chemicL thanks for the excellent work! I can run the unit test on our mcp streamable http service.

But one issue is that our service is running on springframework 5.3.28. And I notice that your WebClientStreamableHttpTransport only work compatible with springframework 6.2.1.

So when I integrate with your code in our service. The below error popped up.
```
java.lang.NoClassDefFoundError: org/springframework/http/HttpStatusCode
```
this is from spring-web-6.2.1.jar

We integrated with HttpClientSseClientTransport before to take care of sse, and the springframework 5.3.28 works fine with the class. Now we would like to upgrade to WebClientStreamableHttpTransport and has this incompatible issue. Is it possible to make it compatible with the old spring version? Thanks!

chemicL commented:
@bugzyz thank you! We are targetting Spring Framework gen 6, unfortunately. However, in your circumstances, once we introduce the JDK HttpClient support for Streamable HTTP, you should be able to use that in combination with older Spring Framework-based code. That work is coming once we finalize the architectural changes covered here. And then you wouldn't need the mcp-spring module dependency. Hope this helps.

noear commented:
Streamable is finally here:) Thanks!

chemicL commented on mcp/src/main/java/io/modelcontextprotocol/client/McpAsyncClient.java:
```diff
@@ -80,24 +80,31 @@ public class McpAsyncClient {
+	};
 
-	private AtomicBoolean initialized = new AtomicBoolean(false);
+	private final AtomicReference<Initialization> initialization = new AtomicReference<>();
```
agreed, thx

chemicL commented on mcp/src/main/java/io/modelcontextprotocol/client/McpAsyncClient.java:
```diff
@@ -234,16 +232,31 @@ public class McpAsyncClient {
 
+	private void handleException(Throwable t) {
+		logger.warn("Handling exception", t);
+		if (t instanceof McpSessionNotFoundException) {
```
Absolutely agreed, thx!

chemicL commented on mcp/src/main/java/io/modelcontextprotocol/client/McpAsyncClient.java:
```diff
@@ -234,16 +232,31 @@ public class McpAsyncClient {
+			if (previous != null) {
+				previous.close();
+			}
+			withSession("re-initializing", result -> Mono.empty()).subscribe();
```
I'll add a comment to clarify. In a nutshell, the argument is the operation to perform once initialization takes place - here we don't do anything, so we provide a `Mono.empty()`.

chemicL commented on mcp/src/main/java/io/modelcontextprotocol/client/McpAsyncClient.java:
```diff
@@ -234,16 +232,31 @@ public class McpAsyncClient {
 	 */
 	public McpSchema.ServerCapabilities getServerCapabilities() {
-		return this.serverCapabilities;
+		Initialization current = this.initialization.get();
```
Thanks for catching this.

chemicL commented on mcp/src/main/java/io/modelcontextprotocol/client/McpAsyncClient.java:
```diff
@@ -326,26 +353,24 @@ public Mono<Void> closeGracefully() {
+		return withSession("by explicit API call", init -> Mono.just(init.get()));
+	}
 
+	private Mono<McpSchema.InitializeResult> doInitialize(McpClientSession session) {
```
went with mcpClientSession to be more precise

chemicL commented on mcp/src/main/java/io/modelcontextprotocol/client/McpAsyncClient.java:
```diff
@@ -355,28 +380,93 @@ public Mono<McpSchema.InitializeResult> initialize() {
+			this.mcpClientSession.set(mcpClientSession);
+		}
+
+		McpClientSession session() {
```
Yep, thx

bugzyz commented:
> @bugzyz thank you! We are targetting Spring Framework gen 6, unfortunately. However, in your circumstances, once we introduce the JDK HttpClient support for Streamable HTTP, you should be able to use that in combination with older Spring Framework-based code. That work is coming once we finalize the architectural changes covered here. And then you wouldn't need the mcp-spring module dependency. Hope this helps.

Thank you @chemicL . Looking forward to the release!

chemicL commented:
Added the documentation, cleaned up. If there are no more remarks, I'll merge this.

The occasional CI failures with STDIO Transport are a result of running `npx -y @modelcontextprotocol/server-everything stdio` within the test which actually first makes a network call to download and install the node package first and than can take several seconds. Potentially we can fix that as a follow-up and run `npm install` for that package in the `@BeforeAll` junit utility and avoid timeouts.

marianogonzalez commented:
@chemicL  this  in response to [this comment](https://github.com/modelcontextprotocol/java-sdk/issues/273#issuecomment-2959952592) in issue #273 .

I think this is definitively a step in the right direction. I think what'd be good for completing this is also adding access to the transport's underlying properties. For instance, it'd be great to have a way of accessing the request's headers and query params. This is important for implementing authorization schemes, custom traceability and other concerns that go beyond what this SDK (or the Spring ecosystem) should implement OOTB (I also think that enabling non-spring apps is very important too)

Also, if you're looking for input on the server side, please see issues #274 and #275 as well.

Thank you!

chemicL commented:
Thanks, @marianogonzalez. We are currently also exploring the mechanisms to allow OAuth2 integration and this work will also dictate some means to inspect the token presence and drive, hopefully transparent, integration throughout the stack. Most preferably, for the other concerns, it would be best to open an issue with a particular need and the intended use cases and scenarios so that we can come up with a solution. Exposing the entire transport-layer headers set makes sense in the context of HTTP, but we can also have a client backed by STDIO-driven local subprocess so some abstractions need to adjust to both cases.
